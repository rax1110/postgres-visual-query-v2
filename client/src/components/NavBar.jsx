import * as PropTypes from 'prop-types';
import React, { Fragment, useState } from 'react';
import {
  Button,
  ButtonDropdown,
  Collapse,
  DropdownMenu,
  DropdownToggle,
  ListGroup,
  ListGroupItem,
  Navbar,
} from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { connect } from 'react-redux';
import _ from 'lodash';
import { translations } from '../utils/translations';
import NavBarMenu from './NavBarMenu';
import NavBarQueryTabs from './NavBarQueryTabs';
import { getCorrectQueryName } from '../utils/getCorrectQueryName';

const NavBar = ({ language, queries }) => {
  const [showNavBarMenu, setShowNavBarMenu] = useState(true);
  const [dropdownOpen, setOpen] = useState(false);
  const [treeElementOpen, setTreeElementOpen] = useState({});

  const toggleDropdown = () => setOpen(!dropdownOpen);

  const handleVisibility = () => {
    setShowNavBarMenu(!showNavBarMenu);
  };

  const mapper = (queriesObj, parentId, lvl) => queriesObj.map((query) => {
    const toggleTreeElement = (e) => {
      const id = e.target.getAttribute('id');

      setTreeElementOpen(state => ({ [id]: !state[id] }));
    };

    const id = `${query.id}-${parentId || 'top'}`.replace(/[^a-zA-Z0-9-_]/g, '');

    return (
      <Fragment key={_.uniqueId('fragment_')}>
        <ListGroupItem
          key={id}
          style={{ zIndex: 0 }}
          className={`${parentId ? `rounded-0 ${lvl ? 'border-bottom-0' : ''}` : ''}`}
        >
          <div style={{ paddingLeft: `${25 * lvl}px` }}>
            {query.subqueries && (
              <Button
                color="link"
                className="pt-0 ml-n3"
                id={id}
                key={id}
                onClick={toggleTreeElement}
              >
                {treeElementOpen[id]
                  ? <FontAwesomeIcon icon="sort-down" className="fa-rotate-270" />
                  : <FontAwesomeIcon icon="sort-down" />}
              </Button>
            )}
            {getCorrectQueryName(language, query.queryName, query.id)}
          </div>
        </ListGroupItem>
        {query.subqueries
        && (
          <Collapse isOpen={treeElementOpen[id]}>
            {mapper(query.subqueries, id, (lvl || 0) + 1)}
          </Collapse>
        )}
      </Fragment>
    );
  });

  return (
    <div className="mb-2">
      <Navbar
        className="m-0 pt-1 pl-2 pb-0"
        color="light"
        light
        expand="sm"
      >
        <ButtonDropdown size="sm" isOpen={dropdownOpen} toggle={toggleDropdown}>
          <DropdownToggle className="btn-sm btn-light btn-outline-secondary" caret>
            <FontAwesomeIcon
              icon="project-diagram"
              size="1x"
              className="pr-2"
              style={{ width: '1.6rem' }}
            />
            {translations[language.code].queryBuilder.queriesH}
          </DropdownToggle>
          <DropdownMenu className="p-0">
            <ListGroup>
              {mapper(queries)}
            </ListGroup>
          </DropdownMenu>
        </ButtonDropdown>
        <span
          className="pl-2 mb-2"
          style={{
            color: '#989ea4',
            fontSize: '1.7rem',
          }}
        >
          |
        </span>
        <div className="col-10 px-1 pl-2 flex d-flex align-items-center flex-wrap">
          <NavBarQueryTabs language={language} />
        </div>
        <div
          className="col-auto pt-3 mr-3 d-flex justify-content-start ml-auto w-auto position-absolute"
          style={{ right: '0', top: '0' }}
        >
          <Button
            onClick={handleVisibility}
            size="sm"
            className="btn-light btn-outline-secondary border-0"
            style={{ backgroundColor: '#f8f9fa' }}
          >
            <FontAwesomeIcon
              icon="angle-down"
              size="2x"
              color="#6c757d"
              className={showNavBarMenu ? 'fa-rotate-180' : ''}
            />
          </Button>
        </div>
      </Navbar>
      {showNavBarMenu && <NavBarMenu language={language} />}
    </div>
  );
};

NavBar.propTypes = {
  language: PropTypes.shape({ code: PropTypes.string }),
  queries: PropTypes.arrayOf(PropTypes.shape({})),
};

const mapStateToProps = (store) => {
  const queries = [...store.queries, store.query].slice()
    .sort((query1, query2) => query1.id - query2.id);
  // .map(query => ({ ...query, subqueries: [{ id: _.uniqueId(), queryName: 'subquery' }] }));

  return ({
    queries,
  });
};

export default connect(mapStateToProps)(NavBar);
