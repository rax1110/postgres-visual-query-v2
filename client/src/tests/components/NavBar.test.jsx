import React from 'react';
import * as renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import NavBar from '../../components/NavBar';

describe('Component: NavBar', () => {
  let store;
  let mockStore;
  let component;
  let props;

  beforeEach(() => {
    props = {
      language: { code: 'eng' },
      query: {
        id: 0,
        queryName: 'Main',
      },
      queries: [
        {
          id: 1,
          queryName: 'Query 1',
        },
        {
          id: 2,
          queryName: 'Query 2',
        },
      ],
    };

    mockStore = configureStore([thunk]);
    store = mockStore(props);
    component = (
      <Provider store={store}>
        <NavBar {...props} />
      </Provider>
    );
  });

  it('renders correctly', () => {
    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('NavBar handleVisibility should toggle NavBarMenu visibility', () => {
    component = mount(component);

    component.find('Button').at(1).simulate('click');

    expect(component.find('NavBarMenu').exists()).toBeTruthy();

    component.find('Button').at(4).simulate('click');

    expect(component.find('NavBarMenu').exists()).toBeFalsy();
  });
});
