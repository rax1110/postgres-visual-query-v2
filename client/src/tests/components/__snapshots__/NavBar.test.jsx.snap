// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: NavBar renders correctly 1`] = `
<div
  className="mb-2"
>
  <nav
    className="m-0 pt-1 pl-2 pb-0 navbar navbar-expand-sm navbar-light bg-light"
  >
    <div
      className="btn-group btn-group-sm"
      onKeyDown={[Function]}
    >
      <button
        aria-expanded={false}
        aria-haspopup={true}
        aria-label={null}
        className="btn-sm btn-light btn-outline-secondary dropdown-toggle btn btn-secondary"
        onClick={[Function]}
        type="button"
      >
        Queries
      </button>
      <div
        aria-hidden={true}
        className="p-0 dropdown-menu"
        role="menu"
        tabIndex="-1"
      >
        <ul
          className="list-group"
        >
          <li
            className="list-group-item"
            style={
              Object {
                "zIndex": 0,
              }
            }
          >
            <div
              style={
                Object {
                  "paddingLeft": "NaNpx",
                }
              }
            >
              Main
            </div>
          </li>
          <li
            className="list-group-item"
            style={
              Object {
                "zIndex": 0,
              }
            }
          >
            <div
              style={
                Object {
                  "paddingLeft": "NaNpx",
                }
              }
            >
              Query 1
            </div>
          </li>
          <li
            className="list-group-item"
            style={
              Object {
                "zIndex": 0,
              }
            }
          >
            <div
              style={
                Object {
                  "paddingLeft": "NaNpx",
                }
              }
            >
              Query 2
            </div>
          </li>
        </ul>
      </div>
    </div>
    <span
      className="pl-2 mb-2"
      style={
        Object {
          "color": "#989ea4",
          "fontSize": "1.7rem",
        }
      }
    >
      |
    </span>
    <div
      className="col-10 px-1 pl-2 flex d-flex align-items-center flex-wrap"
    >
      <div
        className="pr-1 pt-1 pb-1"
      >
        <button
          aria-label={null}
          className="btn-sm btn-secondary btn btn-secondary"
          onClick={[Function]}
          type="button"
          value="Main"
        >
          <a
            className="p-0 d-inline nav-link"
            onClick={[Function]}
          >
            Main
          </a>
        </button>
      </div>
      <div
        className="pr-1 pt-1 pb-1"
      >
        <button
          aria-label={null}
          className="btn-sm btn-light btn-outline-secondary btn btn-secondary"
          onClick={[Function]}
          type="button"
          value="Query 1"
        >
          <a
            className="p-0 d-inline nav-link"
            onClick={[Function]}
          >
            Query 1
          </a>
        </button>
      </div>
      <div
        className="pr-1 pt-1 pb-1"
      >
        <button
          aria-label={null}
          className="btn-sm btn-light btn-outline-secondary btn btn-secondary"
          onClick={[Function]}
          type="button"
          value="Query 2"
        >
          <a
            className="p-0 d-inline nav-link"
            onClick={[Function]}
          >
            Query 2
          </a>
        </button>
      </div>
    </div>
    <div
      className="col-auto pt-3 mr-3 d-flex justify-content-start ml-auto w-auto position-absolute"
      style={
        Object {
          "right": "0",
          "top": "0",
        }
      }
    >
      <button
        aria-label={null}
        className="btn-light btn-outline-secondary border-0 btn btn-secondary btn-sm"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#f8f9fa",
          }
        }
        type="button"
      />
    </div>
  </nav>
  <div
    className="pl-2 align-self-start m-0 pt-1 pb-2 bg-light"
  >
    <div
      className="col-12 pl-0 text-info pt-6"
    >
      <button
        aria-label={null}
        className="btn btn-outline-info btn-sm"
        id="newQuery"
        onClick={[Function]}
        type="button"
      />
       Add query
    </div>
  </div>
</div>
`;
